src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReplacementService.cs

```diff
@@ -1002,25 +1008,261 @@ private async Task ReplaceIndicatorFootnote(Guid footnoteId, TargetableReplaceme
             });
         }
 
-        private async Task ReplaceDataGuidance(
-            Guid releaseId,
-            Guid originalSubject,
-            Guid replacementSubject)
+        private async Task ReplaceReleaseSubject(Guid releaseId,
+            Guid originalSubjectId,
+            Guid replacementSubjectId)
         {
             var originalReleaseSubject = await _statisticsDbContext.ReleaseSubject
-                .AsQueryable()
-                .Where(rs => rs.ReleaseId == releaseId &&
-                             rs.SubjectId == originalSubject)
-                .FirstAsync();
+                .SingleAsync(rs => rs.ReleaseId == releaseId &&
+                                   rs.SubjectId == originalSubjectId);
 
             var replacementReleaseSubject = await _statisticsDbContext.ReleaseSubject
-                .AsQueryable()
-                .Where(rs => rs.ReleaseId == releaseId &&
-                             rs.SubjectId == replacementSubject)
-                .FirstAsync();
+                .SingleAsync(rs => rs.ReleaseId == releaseId &&
+                                   rs.SubjectId == replacementSubjectId);
 
             _statisticsDbContext.Update(replacementReleaseSubject);
+
             replacementReleaseSubject.DataGuidance = originalReleaseSubject.DataGuidance;
+
+            replacementReleaseSubject.FilterSequence =
+                await ReplaceFilterSequence(originalReleaseSubject, replacementReleaseSubject);
+            replacementReleaseSubject.IndicatorSequence =
+                await ReplaceIndicatorSequence(originalReleaseSubject, replacementReleaseSubject);
+        }
+
+        private async Task<List<FilterSequenceEntry>?> ReplaceFilterSequence(ReleaseSubject originalReleaseSubject,
+            ReleaseSubject replacementReleaseSubject)
+        {
+            // If the sequence is undefined then leave it so we continue to fallback to ordering by label alphabetically
+            if (originalReleaseSubject.FilterSequence == null)
+            {
+                return null;
+            }
+
+            var originalFilters = (await _filterRepository.GetFiltersIncludingItems(originalReleaseSubject.SubjectId))
+                .ToDictionary(filter => filter.Name, filter => filter);
+            var replacementFilters =
+                await _filterRepository.GetFiltersIncludingItems(replacementReleaseSubject.SubjectId);
+
+            // Step 1: Create id to replacement-id maps and work out newly added filters, filter groups and filter items
+
+            var filtersMap = new Dictionary<Guid, Guid>();
```

Comment:
Cerrikey!  Good going here.  Just wondering if there's any benefit in hiving some of this code into a helper that could be tested independently of the service?  Not actually that important, just a suggestion when I saw the currently line count of this service!

Reply:
Thanks, it sounds like a good idea to cut down the line count!

I've moved the body of the new replacement code that I added to a new `ReplacementServiceHelper`.

This is unit tested with the service tests taken from my original commit, minus all the setup of the content release, files, database and other dependencies, since it doesn't need any of them now it's independent.

I've also left a much simplified version of the tests in place for the service in `ReplacementServiceTests' to continue exercising the methods. They continue to use the in-memory data and dependencies.

Reply:
Nice one that sounds great!